### PR TITLE
[Feature] Add Maskfeat Support

### PR DESCRIPTION
**Pull request by daidaiershidi**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add maskfeat support.

## Modification

Add MaskGenerator.
Add backbone of maskfeat.
Add head of maskfeat.
Add algorithms of maskfeat.
Add config of maskfeat.
Add Config of MaskFeat.
Update Readme of MaskFeat.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmselfsup/datasets/pipelines/transforms.py [hunk 1]
```diff
@@ -482,3 +482,112 @@ def __repr__(self):
...
+
+    This module is borrowed from
+    https://github.com/facebookresearch/SlowFast/blob/main/slowfast/datasets/transform.py
+    Args:
+        mask_window_size (int): Size of input image. Defaults to 14.
+        mask_ratio (float): The mask ratio of image. Defaults to 0.4.
+        min_num_patches (int): Minimum number of patches that require masking.
+            Defaults to 15.
+        max_num_patches (int, optional): Maximum number of patches that
+            require masking.
```
Comment (tonysy) [inline mmselfsup/datasets/pipelines/transforms.py]: Please add Default value

File: mmselfsup/datasets/pipelines/transforms.py [hunk 2]
```diff
@@ -482,3 +482,112 @@ def __repr__(self):
...
+        min_aspect (int): Minimum aspect of patches. Defaults to 0.3.
+        max_aspect (float, optional): Maximum aspect of patches.
```
Comment (tonysy) [inline mmselfsup/datasets/pipelines/transforms.py:501]: Please add Default value

File: mmselfsup/models/heads/maskfeat_head.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+    Args:
+        embed_dim (int): The dim of the feature before the classifier head.
+        num_classes (int): The total classes. Defaults to 1000.
+    """
+
+    def __init__(self,
+                 embed_dim: int,
+                 num_classes: int = 1000,
+                 label_smooth_val: float = 0.1) -> None:
```
Comment (tonysy) [inline mmselfsup/models/heads/maskfeat_head.py:81]: Please update the docstring for this parameter

File: mmselfsup/models/heads/maskfeat_head.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+@HEADS.register_module()
+class MaskFeatPretrainHead(BaseModule):
+    """Pre-training head for MaskFeat.
+
+    Args:
+        embed_dim (int): The dim of the feature before the classifier head.
+            Defaults to 768.
+        hog_dim (int): The dim of the hog feature. Defaults to 108.
+        reduction (str): Specifies reduction to apply to the output.
```
Comment (tonysy) [inline mmselfsup/models/heads/maskfeat_head.py]: This line is invalid

File: configs/selfsup/_base_/models/maskfeat_vit-base-p16.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+model = dict(
+    type='MaskFeat',
+    backbone=dict(
+        type='MaskFeatViT',
+        arch='b',
+        patch_size=16,
+        drop_path_rate=0,
+    ),
+    head=dict(type='MaskFeatPretrainHead', hog_dim=108),
+    hog_para=dict(nbins=9, pool=8, gaussian_window=16))
```
Comment (tonysy) [inline configs/selfsup/_base_/models/maskfeat_vit-base-p16.py]: please add some comments for hog related parameters, to help the users.

**Review by tonysy — CHANGES_REQUESTED** [review]
As inline comments

File: mmselfsup/models/algorithms/maskfeat.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+
+@ALGORITHMS.register_module()
+class MaskFeat(BaseModel):
+    """MaskFeat.
+
+    Implementation of `Masked Feature Prediction for
+    Self-Supervised Visual Pre-Training <https://arxiv.org/abs/2112.09133>`_.
+    Args:
+        backbone (dict): Config dict for encoder. Defaults to None.
```
Comment (fangyixiao18) [inline mmselfsup/models/algorithms/maskfeat.py]: no default value of `backbone`, `head`, `hog_para`
the docstring of `Defaults to None` can be removed

File: mmselfsup/models/algorithms/maskfeat.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        head (dict): Config dict for loss functions. Defaults to None.
+        hog_para (dict): Config dict for hog layer. Defaults to None.
+            dict['nbins', int]: Number of bin. Defaults to 9.
```
Comment (fangyixiao18) [inline mmselfsup/models/algorithms/maskfeat.py:21]: These comments can be also added to config file

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]

**Review by fangyixiao18 — APPROVED** [review]

File: configs/benchmarks/classification/imagenet/maskfeat_vit-base-p16_ft-8xb512-coslr-100e_in1k.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
+_base_ = [
+    '../_base_/models/vit-base-p16_ft.py',
+    '../_base_/datasets/imagenet.py',
+    '../_base_/schedules/adamw_coslr-100e_in1k.py',
+    '../_base_/default_runtime.py',
+]
+
```
Comment (fangyixiao18) [inline configs/benchmarks/classification/imagenet/maskfeat_vit-base-p16_ft-8xb512-coslr-100e_in1k.py]: add comment `maskfeat fine-tuning setting`

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]

File: docs/zh_cn/model_zoo.md [hunk 9]
```diff
@@ -63,11 +64,12 @@
...
+| -------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ---- | --------- |
+| MAE      | [mae_vit-base-p16_8xb512-coslr-400e_in1k](https://github.com/open-mmlab/mmselfsup/blob/master/configs/selfsup/mae/mae_vit-base-p16_8xb512-coslr-400e_in1k.py)                                          |      | 83.1      |
+| SimMIM   | [simmim_swin-base_16xb128-coslr-100e_in1k-192](https://github.com/open-mmlab/mmselfsup/blob/master/configs/selfsup/simmim/simmim_swin-base_16xb128-coslr-100e_in1k-192.py)                             |      | 82.9      |
+| CAE      | [cae_vit-base-p16_8xb256-fp16-coslr-300e_in1k](https://github.com/open-mmlab/mmselfsup/blob/master/configs/selfsup/cae/cae_vit-base-p16_8xb256-fp16-coslr-300e_in1k.py)                                |      | 83.2      |
+| MaskFeat | [maskfeat_vit-base-p16_8xb256-fp16-coslr-300e_in1k](https://github.com/open-mmlab/mmselfsup/blob/master/configs/benchmarks/classification/imagenet/maskfeat_vit-base-p16_ft-8xb512-coslr-100e_in1k.py) |      | 83.4      |
```
Comment (fangyixiao18) [inline docs/zh_cn/model_zoo.md]:
```suggestion
| MaskFeat | [maskfeat_vit-base-p16_8xb256-fp16-coslr-300e_in1k](https://github.com/open-mmlab/mmselfsup/blob/master/configs/benchmarks/classification/imagenet/maskfeat_vit-base-p16_ft-8xb512-coslr-100e_in1k.py) |      | 83.5      |
```

File: docs/en/model_zoo.md [hunk 10]
```diff
@@ -63,11 +64,12 @@ If not specified, we use linear evaluation setting from [MoCo](http://openaccess
...
+| --------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------- | --------- |
+| MAE       | [mae_vit-base-p16_8xb512-coslr-400e_in1k](https://github.com/open-mmlab/mmselfsup/blob/master/configs/selfsup/mae/mae_vit-base-p16_8xb512-coslr-400e_in1k.py)                                          |         | 83.1      |
+| SimMIM    | [simmim_swin-base_16xb128-coslr-100e_in1k-192](https://github.com/open-mmlab/mmselfsup/blob/master/configs/selfsup/simmim/simmim_swin-base_16xb128-coslr-100e_in1k-192.py)                             |         | 82.9      |
+| CAE       | [cae_vit-base-p16_8xb256-fp16-coslr-300e_in1k](https://github.com/open-mmlab/mmselfsup/blob/master/configs/selfsup/cae/cae_vit-base-p16_8xb256-fp16-coslr-300e_in1k.py)                                |         | 83.2      |
+| MaskFeat  | [maskfeat_vit-base-p16_8xb256-fp16-coslr-300e_in1k](https://github.com/open-mmlab/mmselfsup/blob/master/configs/benchmarks/classification/imagenet/maskfeat_vit-base-p16_ft-8xb512-coslr-100e_in1k.py) |         | 83.4      |
```
Comment (fangyixiao18) [inline docs/en/model_zoo.md]:
```suggestion
| MaskFeat  | [maskfeat_vit-base-p16_8xb256-fp16-coslr-300e_in1k](https://github.com/open-mmlab/mmselfsup/blob/master/configs/benchmarks/classification/imagenet/maskfeat_vit-base-p16_ft-8xb512-coslr-100e_in1k.py) |         | 83.5     |
```

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]

**Review by fangyixiao18 — APPROVED** [review]

File: mmselfsup/datasets/pipelines/transforms.py [hunk 11]
```diff
@@ -482,3 +482,113 @@ def __repr__(self):
         repr_str += f'threshold = {self.threshold}, '
         repr_str += f'prob = {self.prob}'
         return repr_str
+
+
+@PIPELINES.register_module()
+class MaskfeatMaskGenerator(object):
```
Comment (tonysy) [inline mmselfsup/datasets/pipelines/transforms.py]: How about using `MaskFeatureMaskGenerator`? @fangyixiao18 
```suggestion
class MaskFeatMaskGenerator(object):
```

Comment (fangyixiao18) [reply mmselfsup/datasets/pipelines/transforms.py]: I think it is better to use `MaskFeatMaskGenerator`，consisting with algorithm name in our implementation

**Review by tonysy — APPROVED** [review]
Update the name and this pr could be merged.

**Pull request merged**